File: langs/typescript/src/aggregates/aggregateVM3.ts
/**
 * AggregateVM3<VM1, VM2, VM3> — arity-3 aggregate viewmodel.
 *
 * See spec/08-aggregate-vm.md.
 */
import { ComponentVMBase } from "../components/componentVMBase.js";
import { ViewModelType } from "../components/types.js";
import { PropertyChangedMessage } from "../messages/propertyChanged.js";
import type { IMessageHub } from "../services/messageHub.js";
import type { IDispatcher } from "../services/dispatcher.js";
import { BuilderValidationError } from "../builders/exceptions.js";

const SENTINEL = Symbol("not-set");

export class AggregateVM3<
  VM1 extends ComponentVMBase,
  VM2 extends ComponentVMBase,
  VM3 extends ComponentVMBase,
> extends ComponentVMBase {
  readonly #factory1: () => VM1;
  readonly #factory2: () => VM2;
  readonly #factory3: () => VM3;
  #component1: VM1 | null = null;
  #component2: VM2 | null = null;
  #component3: VM3 | null = null;

  constructor(opts: {
    name: string; hint: string; hub: IMessageHub; dispatcher: IDispatcher;
    factory1: () => VM1; factory2: () => VM2; factory3: () => VM3;
  }) {
    super(opts);
    this.#factory1 = opts.factory1; this.#factory2 = opts.factory2; this.#factory3 = opts.factory3;
  }

  get type(): ViewModelType { return ViewModelType.Aggregate; }
  get component1(): VM1 | null { return this.#component1; }
  get component2(): VM2 | null { return this.#component2; }
  get component3(): VM3 | null { return this.#component3; }

  protected override _onConstruct(): void {
    // On Reconstruct, dispose previous slot instances before overwriting
    // so their hub subscriptions and command Subjects don't leak.
    this.#component1?.dispose();
    this.#component2?.dispose();
    this.#component3?.dispose();

    this.#component1 = this.#factory1();
    this._hub.send(PropertyChangedMessage.create(this, this._name, "component1"));
    this._raisePropertyChanged("component1");

    this.#component2 = this.#factory2();
    this._hub.send(PropertyChangedMessage.create(this, this._name, "component2"));
    this._raisePropertyChanged("component2");

    this.#component3 = this.#factory3();
    this._hub.send(PropertyChangedMessage.create(this, this._name, "component3"));
    this._raisePropertyChanged("component3");

    this.#component1.construct();
    this.#component2.construct();
    this.#component3.construct();
  }

  protected override _onDestruct(): void {
    this.#component1?.destruct();
    this.#component2?.destruct();
    this.#component3?.destruct();
  }

  override dispose(): void {
    // Depth-first dispose (LIFE-013): each component slot first, then self.
    this.#component1?.dispose();
    this.#component2?.dispose();
    this.#component3?.dispose();
    super.dispose();
  }

  static builder<
    VM1 extends ComponentVMBase,
    VM2 extends ComponentVMBase,
    VM3 extends ComponentVMBase,
  >(): AggregateVM3Builder<VM1, VM2, VM3> {
    return new AggregateVM3Builder<VM1, VM2, VM3>();
  }
}

export class AggregateVM3Builder<
  VM1 extends ComponentVMBase,
  VM2 extends ComponentVMBase,
  VM3 extends ComponentVMBase,
> {
  #name: string | null = null;
  #hint = "";
  #hub: IMessageHub | null = null;
  #dispatcher: IDispatcher | null = null;
  #factory1: (() => VM1) | typeof SENTINEL = SENTINEL;
  #factory2: (() => VM2) | typeof SENTINEL = SENTINEL;
  #factory3: (() => VM3) | typeof SENTINEL = SENTINEL;

  constructor(from?: AggregateVM3Builder<VM1, VM2, VM3>) {
    if (from) {
      this.#name = from.#name; this.#hint = from.#hint;
      this.#hub = from.#hub; this.#dispatcher = from.#dispatcher;
      this.#factory1 = from.#factory1; this.#factory2 = from.#factory2;
      this.#factory3 = from.#factory3;
    }
  }

  name(v: string): AggregateVM3Builder<VM1, VM2, VM3> { const b = new AggregateVM3Builder<VM1, VM2, VM3>(this); b.#name = v; return b; }
  hint(v: string): AggregateVM3Builder<VM1, VM2, VM3> { const b = new AggregateVM3Builder<VM1, VM2, VM3>(this); b.#hint = v; return b; }
  services(hub: IMessageHub, disp: IDispatcher): AggregateVM3Builder<VM1, VM2, VM3> { const b = new AggregateVM3Builder<VM1, VM2, VM3>(this); b.#hub = hub; b.#dispatcher = disp; return b; }
  component1(f: () => VM1): AggregateVM3Builder<VM1, VM2, VM3> { const b = new AggregateVM3Builder<VM1, VM2, VM3>(this); b.#factory1 = f; return b; }
  component2(f: () => VM2): AggregateVM3Builder<VM1, VM2, VM3> { const b = new AggregateVM3Builder<VM1, VM2, VM3>(this); b.#factory2 = f; return b; }
  component3(f: () => VM3): AggregateVM3Builder<VM1, VM2, VM3> { const b = new AggregateVM3Builder<VM1, VM2, VM3>(this); b.#factory3 = f; return b; }

  build(): AggregateVM3<VM1, VM2, VM3> {
    if (this.#name === null) throw new BuilderValidationError("name");
    if (this.#hub === null || this.#dispatcher === null) throw new BuilderValidationError("services");
    if (this.#factory1 === SENTINEL) throw new BuilderValidationError("component1");
    if (this.#factory2 === SENTINEL) throw new BuilderValidationError("component2");
    if (this.#factory3 === SENTINEL) throw new BuilderValidationError("component3");
    const factory1 = this.#factory1;
    const factory2 = this.#factory2;
    const factory3 = this.#factory3;
    return new AggregateVM3<VM1, VM2, VM3>({
      name: this.#name, hint: this.#hint, hub: this.#hub, dispatcher: this.#dispatcher,
      factory1, factory2, factory3,
    });
  }
}

File: langs/csharp/src/VMx/Aggregates/AggregateVM3.cs
using VMx.Builders;
using VMx.Components;
using VMx.Services;

namespace VMx.Aggregates;

/// <summary>
/// Arity-3 aggregate viewmodel. A fixed tuple of three heterogeneous component VMs.
/// Component slots are populated lazily on construct() by invoking factories
/// provided via the nested builder.
///
/// See spec/08-aggregate-vm.md and ADR-0007.
/// </summary>
/// <typeparam name="VM1">Type of the first component.</typeparam>
/// <typeparam name="VM2">Type of the second component.</typeparam>
/// <typeparam name="VM3">Type of the third component.</typeparam>
public sealed class AggregateVM3<VM1, VM2, VM3> : ComponentVMBase, IAggregateVM3<VM1, VM2, VM3>, IAggregateSlots
    where VM1 : class, IComponentVM
    where VM2 : class, IComponentVM
    where VM3 : class, IComponentVM
{
    private readonly Func<VM1> _factory1;
    private readonly Func<VM2> _factory2;
    private readonly Func<VM3> _factory3;
    private VM1? _component1;
    private VM2? _component2;
    private VM3? _component3;

    IEnumerable<IComponentVM> IAggregateSlots.EnumerateSlots()
    {
        if (_component1 is { } c1) yield return c1;
        if (_component2 is { } c2) yield return c2;
        if (_component3 is { } c3) yield return c3;
    }

    // ── IAggregateVM3<VM1, VM2, VM3> ────────────────────────────────────────

    /// <inheritdoc/>
    public VM1? Component1 => _component1;

    /// <inheritdoc/>
    public VM2? Component2 => _component2;

    /// <inheritdoc/>
    public VM3? Component3 => _component3;

    // ── IComponentVM.Type ───────────────────────────────────────────────────

    /// <inheritdoc/>
    public override ViewModelType Type => ViewModelType.Aggregate;

    // ── Constructor ─────────────────────────────────────────────────────────

    private AggregateVM3(
        string name,
        string hint,
        IMessageHub hub,
        IDispatcher dispatcher,
        Func<VM1> factory1,
        Func<VM2> factory2,
        Func<VM3> factory3)
        : base(name, hint, hub, dispatcher, onConstruct: null, onDestruct: null)
    {
        _factory1 = factory1;
        _factory2 = factory2;
        _factory3 = factory3;
    }

    // ── Lifecycle overrides ─────────────────────────────────────────────────

    /// <inheritdoc/>
    protected override void OnConstruct()
    {
        // On Reconstruct, dispose previous slot instances before overwriting
        // so their hub subscriptions and command Subjects don't leak.
        _component1?.Dispose();
        _component2?.Dispose();
        _component3?.Dispose();

        _component1 = _factory1();
        NotifyPropertyChanged(nameof(Component1));

        _component2 = _factory2();
        NotifyPropertyChanged(nameof(Component2));

        _component3 = _factory3();
        NotifyPropertyChanged(nameof(Component3));

        CompleteLifecycleHookAfter(TransitionChildrenAsync(
            [_component1, _component2, _component3], construct: true));
    }

    /// <inheritdoc/>
    protected override void OnDestruct()
    {
        CompleteLifecycleHookAfter(TransitionChildrenAsync(
            new IComponentVM?[] { _component1, _component2, _component3 }
                .OfType<IComponentVM>(),
            construct: false));
    }

    /// <summary>
    /// Dispose cascade (LIFE-013): dispose each component slot depth-first, then self.
    /// </summary>
    public override void Dispose()
    {
        var firstError = DisposeChildren([_component1, _component2, _component3]);
        try { base.Dispose(); }
        catch (Exception error)
        {
            firstError ??= System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(error);
        }
        firstError?.Throw();
    }

    // ── Builder factory ─────────────────────────────────────────────────────

    /// <summary>Returns a new empty builder for <see cref="AggregateVM3{VM1,VM2,VM3}"/>.</summary>
    public static AggregateVM3Builder Builder() => new();

    // ── Nested builder ──────────────────────────────────────────────────────

    /// <summary>
    /// Immutable fluent builder for <see cref="AggregateVM3{VM1,VM2,VM3}"/>.
    /// Each setter returns a NEW builder instance (BLD-001).
    /// </summary>
    public sealed class AggregateVM3Builder
    {
        private readonly string? _name;
        private readonly string _hint;
        private readonly IMessageHub? _hub;
        private readonly IDispatcher? _dispatcher;
        private readonly Func<VM1>? _factory1;
        private readonly Func<VM2>? _factory2;
        private readonly Func<VM3>? _factory3;

        internal AggregateVM3Builder()
        {
            _hint = "";
        }

        private AggregateVM3Builder(
            string? name,
            string hint,
            IMessageHub? hub,
            IDispatcher? dispatcher,
            Func<VM1>? factory1,
            Func<VM2>? factory2,
            Func<VM3>? factory3)
        {
            _name = name;
            _hint = hint;
            _hub = hub;
            _dispatcher = dispatcher;
            _factory1 = factory1;
            _factory2 = factory2;
            _factory3 = factory3;
        }

        /// <summary>Sets the required Name field.</summary>
        public AggregateVM3Builder Name(string name)
            => new(name, _hint, _hub, _dispatcher, _factory1, _factory2, _factory3);

        /// <summary>Sets the optional Hint field (default: empty string).</summary>
        public AggregateVM3Builder Hint(string hint)
            => new(_name, hint, _hub, _dispatcher, _factory1, _factory2, _factory3);

        /// <summary>Sets the required Services (hub + dispatcher).</summary>
        public AggregateVM3Builder Services(IMessageHub hub, IDispatcher dispatcher)
            => new(_name, _hint, hub, dispatcher, _factory1, _factory2, _factory3);

        /// <summary>Sets the required factory for Component1.</summary>
        public AggregateVM3Builder Component1(Func<VM1> factory)
            => new(_name, _hint, _hub, _dispatcher, factory, _factory2, _factory3);

        /// <summary>Sets the required factory for Component2.</summary>
        public AggregateVM3Builder Component2(Func<VM2> factory)
            => new(_name, _hint, _hub, _dispatcher, _factory1, factory, _factory3);

        /// <summary>Sets the required factory for Component3.</summary>
        public AggregateVM3Builder Component3(Func<VM3> factory)
            => new(_name, _hint, _hub, _dispatcher, _factory1, _factory2, factory);

        /// <summary>
        /// Validates required fields and constructs a <see cref="AggregateVM3{VM1,VM2,VM3}"/>.
        /// Throws <see cref="BuilderValidationException"/> if a required field is missing.
        /// </summary>
        public AggregateVM3<VM1, VM2, VM3> Build()
        {
            BuilderValidationException.Require(_name, "Name");
            BuilderValidationException.Require(_hub, "Hub");
            BuilderValidationException.Require(_dispatcher, "Dispatcher");
            BuilderValidationException.Require(_factory1, "Component1");
            BuilderValidationException.Require(_factory2, "Component2");
            BuilderValidationException.Require(_factory3, "Component3");

            return new AggregateVM3<VM1, VM2, VM3>(_name, _hint, _hub, _dispatcher, _factory1, _factory2, _factory3);
        }
    }
}
